### ISSUE TITLE
Convert min-iterations convergence measure to plain tag

### ISSUE THREAD

**Issue by uekerman** [precice/precice#800]
Currently the min-iterations convergence measure needs to be associated to some data even though there is no real connection.
```xml
<min-iteration-convergence-measure data="CrossSectionLength" mesh="Structure_Nodes" min-iterations="2"/>
```

This cannot be fixed "directly" on the configuration level as convergence measures enforce having data:

https://github.com/precice/precice/blob/6437f8e43710818d6ff96074d16e93fca8604088/src/cplscheme/impl/ConvergenceMeasure.hpp#L36-L38

I suggest to convert the convergence measure to a simple tag of `coupling-scheme`, similar to `max-iterations`
```xml
<coupling-scheme:serial-implicit>
      <participants first="FLUID" second="STRUCTURE"/>
      ...
      <max-iterations value="40"/>
      <min-iterations value="5"/>
      ...
</coupling-scheme:serial-implicit>
```

Drawback of this is that then "min iterations" can no longer be combined with `suffices="true"` to any other convergence measure. In my opinion this loss, however, is acceptable. I am not sure if anybody uses that feature anyway.

Should we do this?

**Comment by fsimonis** [comment]
The tag and the convergence measure slightly differ in meaning.

Given the tags, we can define an interval in which the convergence measures are used to stop the iterative process.

For the interval `[5, 40]` one could write:
```xml
<coupling-scheme:serial-implicit>
      <min-iterations value="5"/>
      <max-iterations value="40"/>
</coupling-scheme:serial-implicit>
```
Before the minimum iteration is reached, all measures are ignored.
After the maximum iteration is reached, unconverged strict convergence measures will stop the simulation.

## Covered use-case: hard stop and continue
Once combined with `suffice="true"`, the convergence measure defines an iteration count at which the time window stops (as long as there are no strict cms) and reaches convergence.
This can be expressed in the configuration by setting all convergence measures to not-strict and using `<max-iterations/>` to specify the maximum iterations.

## Not covered use-case: soft minimum
```xml
<coupling-scheme:serial-implicit>
      <absolute-convergence-measure suffices="true" ... />
      <min-iteration-convergence-measure min-iterations="4"/>
      <relative-convergence-measure ... />
</coupling-scheme:serial-implicit>
```

The minimum iterations convergence measure offers a soft minimum for the iterations. The absolute cm stops the simulation if it converges, otherwise the simulation will stop after the relative cm has converged and 4 total iterations have passed.
This functionality is similar to the proposed `min-iterations` tag, but it would force the simulation to continue until a given amount of iterations. One could implement a "weak" version of the minimum, which ends if a `suffices` cm reached convergence before the `min-iterations` was reached.

**Comment by uekerman** [comment]
Exactly!

> One could implement a "weak" version of the minimum, which ends if a suffices cm reached convergence before the min-iterations was reached.

Yes, that would be possible. I think though that this use case is too rare. I suggest to just drop this case and make `min-iterations` always a hard criterion.

**Comment by fsimonis** [comment]
@uekerman Many tests and our solverdummes use the `min-iteration-convergence` measure as a simple method to "run until". This also affects many time tests written by @BenjaminRodenberg.

When we convert the measure to the `min-iterations` tag, then these cases don't have a convergence measure, which isn't allowed.

What should we do in those cases?

**Comment by BenjaminRodenberg** [comment]
> When we convert the measure to the `min-iterations` tag, then these cases don't have a convergence measure, which isn't allowed.

Removing the strict requirement for having a convergence measure would be one possible solution here.

* If there is no convergence measure, we don't care about the coupling error. I think this would also be consistent with how our configuration often works: If there is no acceleration, we do not apply acceleration and do a plain fixed point iteration. If there is no mapping, we do not perform a mapping. 
* Looking at the current name `min-iterations-convergence-measure` this is also a bit counter-intuitive: We are actually not measuring convergence here. It's rather a more general stopping criterion.

One thing we would need to take care of: If no convergence measure is provided, `max-iterations` does have any reasonable effect. Do we want to catch this? Additionally, we should require `min-iterations <= max-iterations`. Finally, what happens, if we do not provide `min-iterations`, `max-iterations` or `*-convergence-measure`?

Alternative suggestion: Why not make `min-iterations` and `max-iterations` attributes of implicit coupling schemes? If we use reasonable default values (`min-iterations="1"`, `max-iterations="inf"`), this simplifies the configuration and avoids misconfiguration by design (e.g. no duplicate and contradictory `max-iterations`/`min-iterations` tags possible)

**Comment by fsimonis** [comment]
> Removing the strict requirement for having a convergence measure would be one possible solution here.

In that case, the scheme will iterate exactly `min-iterations` times and then advance to the next time window.

To my understanding, the constraints, and warnings should then be:
* `min-iterations > 0 && min-iterations <= max-iterations`
* `min-iterations == 1 && no convergence measure` _Warning: The scheme will always iterate exactly once, use an explicit scheme instead._
* `max-iterations < inf && no convergence measure` _Warning: Without convergence measures, the scheme will iterate exactly 42 times (min-iterations)._

> Finally, what happens, if we do not provide min-iterations, max-iterations or *-convergence-measure?

Would the above warning be enough?
_Warning: The scheme will always iterate exactly once, use an explicit scheme instead._


> Alternative suggestion: Why not make min-iterations and max-iterations attributes of implicit coupling schemes?

This rational is true for more tags in the coupling-schemes though.
* max-time
* max-time-windows
* first-participant
* second-participant

I think this clean up belongs into a separate issue. The big questions are: what do we want to clean up and do we want to do this for v3 or v4.

**Comment by BenjaminRodenberg** [comment]
@fsimonis Generally, everything above sounds good to me. Just some quick comments:

> > Finally, what happens, if we do not provide min-iterations, max-iterations or *-convergence-measure?
> 
> Would the above warning be enough? _Warning: The scheme will always iterate exactly once, use an explicit scheme instead._

Alternative: Since the default for `max-iterations` would be `inf`, if nothing is provided, we should consequently iterate forever. Does not make a lot of sense at the moment, but I'm thinking about future possibilities like allowing the user to call `finalize()` inside of the simulation loop (https://github.com/precice/precice/issues/1118).

> I think this clean up belongs into a separate issue. The big questions are: what do we want to clean up and do we want to do this for v3 or v4.

True. Let's discuss this in a different issue. Still: We might also start with `min-iterations` and `max-iterations` with v3, get some experience, and then do the big redesign in v4. If we learn that using attributes is a terrible idea, we could still move back quite easily, by re-introducing the tag and deprecating the attribute. (same is true for the other direction)

**Comment by fsimonis** [comment]
> Alternative: Since the default for max-iterations would be inf, if nothing is provided, we should consequently iterate forever. Does not make a lot of sense at the moment, but I'm thinking about future possibilities like allowing the user to call finalize() inside of the simulation loop (https://github.com/precice/precice/issues/1118).

I think the easiest to understand and develop is:

tag | default | behaviour
---|---|---
`min-iterations` | 1 | allows convergence after n iterations
`max-iterations` | inf | ignores convergence and moves to next time window after n iterations

Not defining a convergence measure means: I iterate `min-iterations` times and move to the next time window.

To implement #1118, there is some kind of additional synchronization required. Possible with current convergence measures and the above config would be #1678.

**Comment by uekerman** [comment]
Sounds all good. Let's do it. But let's not make attributes yet. Let's only break things if we see a clear advantage, not for getting experience :grin:

**Comment by uekerman** [comment]
> Not defining a convergence measure means: I iterate min-iterations times and move to the next time window.

Not sure about this one actually. I would say if there is no convergence measure defined, we do max-iterations. And error if it is `inf`. (This would mean all tests that currently have `min-iterations-c-m`, need to get a `max-iterations`.)

**Comment by fsimonis** [comment]
> > Not defining a convergence measure means: I iterate min-iterations times and move to the next time window.
> 
> Not sure about this one actually. I would say if there is no convergence measure defined, we do max-iterations. And error if it is `inf`. (This would mean all tests that currently have `min-iterations-c-m`, need to get a `max-iterations`.)

Currently, the situation is a bit tricky:

* `min-iterations-c-m` maps to `min-iterations` as we require n iterations before we can converge.
* `min-iterations-c-m suffices="true"` maps to `max-iterations` as we converge as soon as we reach the iteration count.

The open question is:

Do we regard _no convergence_ measure as _always converged_ or _never converged_?


So if we use `min-iterations=1` and `max-iterations=inf` as defaults, what behaviour do we expect for:

### Nothing

```xml
 <coupling-scheme:serial-implicit> 
   <participants first="A" second="B" /> 
   <max-time-windows value="2" /> 
   <time-window-size value="1.0" />
   <exchange ... />
 </coupling-scheme:serial-implicit> 
```

FS: Info: the scheme always iterates exactly once, use an explicit scheme instead.

### Defined Min

```xml
 <coupling-scheme:serial-implicit> 
   <participants first="A" second="B" /> 
   <max-time-windows value="2" /> 
   <time-window-size value="1.0" />
   <exchange ... />
   <min-iterations value="2" />
 </coupling-scheme:serial-implicit> 
```

FS: Info: the scheme always iterates exactly 2 times, you may want to define convergence measures.

### Defined Max

```xml
 <coupling-scheme:serial-implicit> 
   <participants first="A" second="B" /> 
   <max-time-windows value="2" /> 
   <time-window-size value="1.0" />
   <exchange ... />
   <max-iterations value="10" />
 </coupling-scheme:serial-implicit> 
```

FS: Warning/Error: defining max-iterations without convergence measures has no effect.

### Defined Min and Max

```xml
 <coupling-scheme:serial-implicit> 
   <participants first="A" second="B" /> 
   <max-time-windows value="2" /> 
   <time-window-size value="1.0" />
   <exchange ... />
   <min-iterations value="2" />
   <max-iterations value="10" />
 </coupling-scheme:serial-implicit> 
```

FS:
* Info: the scheme always iterates exactly 2 times, you may want to define convergence measures.
* Warning/Error: defining max-iterations without convergence measures has no effect.

**Comment by BenjaminRodenberg** [comment]
> Do we regard _no convergence_ measure as _always converged_ or _never converged_?

or in code

a)

```python
while(
  it < min_it or 
  (
    it < max_it and 
    any([not cm.converged() for cm in cms])  # false, if cms is empty
  )):
  # iterate until it == min_it, if cms is empty
```
or

b)
```python
while(
  it < min_it or 
  (
    it < max_it and 
    not all([cm.converged() for cm in cms])  # true, if cms is empty
  )):
  # iterate until it == max_it, if cms is empty
```

**Comment by BenjaminRodenberg** [comment]
> No convergence measure means no convergence and the scheme runs until max-iterations are reached.

I just realized that this is not ideal when reviewing #1841. The following configuration for the solverdummy looks unnecessarily complicated to me:

```diff
<coupling-scheme:serial-implicit>
  <participants first="SolverOne" second="SolverTwo" />
  <max-time-windows value="2" />
  <time-window-size value="1.0" />
  <max-iterations value="2" />
+ <min-iterations value="2" />
- <min-iteration-convergence-measure min-iterations="5" data="Data-One" mesh="SolverOne-Mesh" />
  ...
 </coupling-scheme:serial-implicit>
```

In the end the goal is just to tell preCICE "do exactly 2 iterations". Additionally, having to provide the same number twice is also error-prone and can lead to strange situations, like seen in this config (Why is `min-iterations="5"` in the old version?).

Why not do it differently? Not "No convergence measure means no convergence and the scheme runs until max-iterations are reached.", but instead "No convergence measure means nothing to worry about and scheme runs until min-iterations are reached".

This would lead to the following, in my opinion more intuitive, config:

```
<coupling-scheme:serial-implicit>
  <participants first="SolverOne" second="SolverTwo" />
  <max-time-windows value="2" />
  <time-window-size value="1.0" />
  <min-iterations value="2" />
  ...
 </coupling-scheme:serial-implicit>
```

The situation [defined min and max without convergence measure](https://github.com/precice/precice/issues/800#issuecomment-1770825738) is still tricky, but I think this is always the case, because the user is obviously doing something wrong here. In any case the user over specifies the config and providing some kind of feedback to the user makes sense.

**Comment by fsimonis** [comment]
> The situation defined min and max without convergence measure is still tricky, but I think this is always the case, because the user is obviously doing something wrong here.

Exactly, this is a leftover from a previous version of the PR. `min-iterations` has no effect here and should result in a warning.

I would prefer not to raise an error in case `min-iterations` is defined without convergence measures, as this could be an unnecessary nuisance when a user disables all convergence measures during debugging.

**Comment by BenjaminRodenberg** [comment]
> Why not do it differently? Not "No convergence measure means no convergence and the scheme runs until max-iterations are reached.", but instead "No convergence measure means nothing to worry about and scheme runs until min-iterations are reached".

Most important part in my opinion is this here. Any opinions on this change? @uekerman and @fsimonis

**Comment by fsimonis** [comment]
I don't think that the one is easier to understand than the other. It's a pure matter of definition.

Assuming that no convergence measure means "never converge" is slightly easier to implement and check for. This is why I used it in the PR.

**Comment by BenjaminRodenberg** [comment]
I think being able to remove all the `max-iteration` tags from the test configs in https://github.com/precice/precice/pull/1841 is an argument in favor for doing it differently - especially if there is no strong reason for picking one solution over the other. Slightly easier to implement does not sound very strong to me.

**Comment by uekerman** [comment]
`min-iterations` should really be "always do at least X iterations". Not that sth is supposed to happen after X iterations.

> The following configuration for the solverdummy looks unnecessarily complicated to me:

You can remove the min iterations there.

```xml
<coupling-scheme:serial-implicit>
  <participants first="SolverOne" second="SolverTwo" />
  <max-time-windows value="2" />
  <time-window-size value="1.0" />
  ...
 </coupling-scheme:serial-implicit>
```

**Comment by fsimonis** [comment]
> I think being able to remove all the max-iteration tags from the test configs in https://github.com/precice/precice/pull/1841 is an argument in favor for doing it differently

Either we rely on `min-iterations` or `max-iterations` in tests that don't use a convergence measure.
Relying on some default can lead to problems in the future, so I would definitely define the one or the other. So, there is no difference. Not even in the amount of characters.

From the user-perspective I think that iterating to `max-iterations` may be the more helpful here.

Let's assume a user configured an implicit coupling scheme with a min and max count of iterations. Something fishy is happening regarding the convergence checks after min iterations are reached. Then in the debugging process the user can turn off the convergence measures and the implicit scheme now run for the full max-iterations. If the scheme stops at min-iterations, then the user needs to set `min-iterations` to `max-iterations` in order to force the scheme to iterate into the interval where convergence measures would have an impact.

**Comment by BenjaminRodenberg** [comment]
So at least we agree that using both `min-iterations` and `max-iterations` when trying to tell preCICE "Do exactly X iterations" is too much.

I'm arguing one last time for `min-iterations`, but if there are 2 votes against one, I can live with any solution and we should move in either of the two directions.

---

I understand the debugging argument from @fsimonis, but this sounds to me more like a nice-to-have and not really like the argument we should base our decision on. I would also say a valid way to achieve the same is to leave the convergence schemes as they are and just increase `min-iterations` to the value of `max-iterations`. This also bypasses the convergence measures. Maybe this approach is even better, because the convergence schemes are more complicated I would rather not delete them from the config (I know, you can also comment out).

Looking at the proposed config by @uekerman, I still don't really think that asking for `max-iterations` sounds like the amount of iterations you always do, if there is no convergence measure. "Do max iterations" means "Don't do more", "Do mean iterations" means "Do at least". 

One final argument: I think using `min-iterations` instead of `max-iterations` makes the transition for our users easier and is also easier to communicate for us.

**Use `max-iterations="X"` to tell preCICE "do exactly X iterations"**

Remove `min-iteration-convergence-measure` and use the value used in `min-iterations="X"` for `<max-iterations value="X" />`. (Good point: We don't have to tell users about the existence of the new `min-iterations` tag at all.)

Assuming the user did everything in a consistent manner before, this means just deleting `min-iteration-convergence-measure`:

```diff
<coupling-scheme:serial-implicit>
  <participants first="SolverOne" second="SolverTwo" />
  <max-iterations value="2" />
- <min-iteration-convergence-measure min-iterations="2" data="Data-One" mesh="SolverOne-Mesh" />
  ...
</coupling-scheme:serial-implicit>
```

But things can become more complicated (pre https://github.com/precice/precice/commit/0d36758ebb45ae1a3b7dc03368c10104d85ffdbe):

```diff
<coupling-scheme:serial-implicit>
  <participants first="SolverOne" second="SolverTwo" />
+ <max-iterations value="2" />
- <max-iterations value="5" />
- <min-iteration-convergence-measure min-iterations="2" data="Data-One" mesh="SolverOne-Mesh" />
  ...
</coupling-scheme:serial-implicit>
```

**Use `min-iterations="X"` to tell preCICE "do exactly X iterations"**

Remove `min-iteration-convergence-measure` and use the value used in `min-iterations="X"` for the new tag `<min-iterations value="X" />`.

```diff
<coupling-scheme:serial-implicit>
  <participants first="SolverOne" second="SolverTwo" />
  <max-iterations value="2" />
- <min-iteration-convergence-measure min-iterations="2" data="Data-One" mesh="SolverOne-Mesh" />
+ <min-iterations value ="2" />
  ...
</coupling-scheme:serial-implicit>
```

For the pre https://github.com/precice/precice/commit/0d36758ebb45ae1a3b7dc03368c10104d85ffdbe situation, because the removal of `max-iterations` has no effect.

```diff
<coupling-scheme:serial-implicit>
  <participants first="SolverOne" second="SolverTwo" />
  <max-iterations value="5" />
- <min-iteration-convergence-measure min-iterations="2" data="Data-One" mesh="SolverOne-Mesh" />
+ <min-iterations value ="2" />
  ...
</coupling-scheme:serial-implicit>
```

We can tell the user already here that `max-iterations` is optional in this case, but we can solve this with an info or warning message independent from this whole process as well.

**Comment by uekerman** [comment]
I don't think that many users use `min-iterations-c-m`. This option is mainly for integration tests. Thus, porting is no argument in my opinion.

Let's not use more time here, this is really an unimportant topic. I vote that `min-iterations` should only be the minimum and not say anything about the maximum, thus

```xml
<coupling-scheme:serial-implicit>
  <participants first="SolverOne" second="SolverTwo" />
  <max-time-windows value="2" />
  <time-window-size value="1.0" />
  <min-iterations value="2" />
  ...
 </coupling-scheme:serial-implicit>
```

will iterate forever (i.e. warning or error)

and

```xml
<coupling-scheme:serial-implicit>
  <participants first="SolverOne" second="SolverTwo" />
  <max-time-windows value="2" />
  <time-window-size value="1.0" />
  <max-iterations value="2" />
  ...
 </coupling-scheme:serial-implicit>
```
does exactly 2 iterations.
